internal/luks2/metadata.go

```diff
@@ -0,0 +1,874 @@
+	}
+
+	h := csumHash.New()
+
+	// Hash the binary header without the checksum
+	hdrTmp := hdr
+	hdrTmp.Csum = [64]byte{}
+	if err := binary.Write(h, binary.BigEndian, &hdrTmp); err != nil {
+		return nil, nil, xerrors.Errorf("cannot calculate checksum, error serializing header: %w", err)
+	}
+
+	// Hash the JSON metadata area, keeping a copy of the hashed metadata in memory
+	var jsonBuffer bytes.Buffer
+	tr := io.TeeReader(r, &jsonBuffer)
+	if _, err := io.CopyN(h, tr, int64(hdr.HdrSize)-int64(binary.Size(hdr))); err != nil {
+		return nil, nil, xerrors.Errorf("cannot calculate checksum, error reading JSON metadata: %w", err)
+	}
+
+	if !bytes.Equal(h.Sum(nil), hdr.Csum[0:csumHash.Size()]) {
+		return nil, nil, errors.New("invalid header checksum")
+	}
+
+	// Return the binary header and the memory buffer containing the verified JSON metadata
+	return &hdr, jsonBuffer.Bytes(), nil
+}
+
+// DecodeHdr will decode the LUKS header at the specified path. The path can either be a block device
+// or file containing a LUKS2 volume with an integral header, or it can be a detached header file.
+// Data is interpreted in accordance with the LUKS2 On-Disk Format specification
+// (https://gitlab.com/cryptsetup/LUKS2-docs/blob/master/luks2_doc_wip.pdf).
+//
+// This function will verify the checksum of both the primary and secondary headers if found, and will
+// return the decoded form of one of the headers according to the following rules:
+//  - If both headers have valid checksums and the same sequence ID, return the primary header.
+//  - If both headers have valid checksums but different sequence IDs, return the newest header.
+//  - If only one header has a valid checksum, return that header.
+//
+// libcryptsetup performs some additional validation of the JSON metadata from both the primary
+// and secondary headers, and rejects a header if the JSON metadata isn't correctly formed. We don't
+// duplicate that logic here - we assume that anything that modifies the LUKS2 headers (which should
+// only be libcryptsetup) will write well-formed JSON metadata. Corruption of the JSON metadata outside
+// of modifications by libcryptsetup will be detected by the checksum verification.
+//
+// Note that this function does not attempt recovery of either header in the event that one of the
+// headers is not valid - we leave this to libcryptsetup, which happens automatically on any
+// cryptsetup or systemd-cryptsetup invocation. This package does not directly perform any
+// modifications to the header.
+//
+// This function requires an advisory shared lock on the LUKS container associated with the
+// specified path. If the mode parameter is LockModeBlocking, this function will block until the
+// lock can be obtained. If the mode parameter is LockModeNonBlocking, a wrapped syscall.Errno
+// error with the value of syscall.EWOULDBLOCK will be returned if the lock can not be obtained.
+func DecodeHdr(path string, lockMode LockMode) (*HdrInfo, error) {
```

Comment:
is the contraction Hdr a naming preferred by LUKS itself?  because otherwise I would call this DecodeHeader.

Same for other uses of Hdr.

Reply:
also is the "Decode" part also naming from LUKS?  because "ReadHeader" would be a more natural/clearer name for this

Reply:
The LUKS spec does use the abbreviation "hdr" in various places - I've changed DecodeHdr to ReadHeader, HdrInfo to HeaderInfo, but I've kept the "Hdr" abbreviation for the un-exported representation of the binary header.